header-validator/src/validate-json.ts

```diff
@@ -26,54 +26,94 @@ class Context extends context.Context {
   }
 }
 
-type FieldCheck = (ctx: Context, obj: JsonDict, key: string) => Maybe<any>
-
-type FieldChecks = Record<string, FieldCheck>
+type StructFields<T extends object> = {
+  [K in keyof T]-?: CtxFunc<JsonDict, Maybe<T[K]>>
+}
 
-function validate(
+function struct<T extends object>(
   ctx: Context,
-  obj: Json,
-  checks: FieldChecks
-): Maybe<JsonDict> {
-  return object(ctx, obj).filter((d) => {
+  d: Json,
+  fields: StructFields<T>
+): Maybe<T> {
+  return object(ctx, d).flatMap((d) => {
+    let t: Partial<T> = {}
+
     let ok = true
-    Object.entries(checks).forEach(([key, check]) => {
+    for (const prop in fields) {
       let itemOk = false
-      ctx.scope(key, () => check(ctx, d, key).peek(() => (itemOk = true)))
+      fields[prop](ctx, d).peek((v) => {
+        itemOk = true
+        t[prop] = v
+      })
       ok = ok && itemOk
-    })
+    }
 
-    Object.keys(d).forEach((key) => {
-      if (!(key in checks)) {
-        ctx.scope(key, () => ctx.warning('unknown field'))
-      }
-    })
+    for (const key in d) {
+      ctx.scope(key, () => ctx.warning('unknown field'))
+    }
 
-    return ok
+    return ok ? new Some(t as T) : None
   })
 }
 
 type CtxFunc<I, O> = (ctx: Context, i: I) => O
 
 export type ValueCheck = CtxFunc<Json, Maybe<any>>
 
-function field(f: ValueCheck, required: boolean): FieldCheck {
-  return (ctx, obj, key) => {
-    const v = obj[key]
-    if (v === undefined) {
-      if (required) {
-        ctx.error('required')
-        return None
+function field<T>(
+  name: string,
+  f: CtxFunc<Json, Maybe<T>>,
+  valueIfAbsent?: T
+): CtxFunc<JsonDict, Maybe<T>> {
+  return (ctx: Context, d: JsonDict): Maybe<T> =>
+    ctx.scope(name, () => {
+      const v = d[name]
+      if (v === undefined) {
+        if (valueIfAbsent === undefined) {
+          ctx.error('required')
+          return None
+        }
+        return new Some(valueIfAbsent)
+      }
+      delete d[name]
+      return f(ctx, v)
+    })
+}
+
+type Exclusive<T> = {
+  [key: string]: CtxFunc<Json, Maybe<T>>
+}
+
+function exclusive<T>(
+  x: Exclusive<T>,
+  valueIfAbsent: T
+): CtxFunc<JsonDict, Maybe<T>> {
+  return (ctx: Context, d: JsonDict): Maybe<T> => {
+    const found: string[] = []
+    let v: Maybe<T> = None
+
+    for (const [key, f] of Object.entries(x)) {
+      const j = d[key]
+      if (j !== undefined) {
+        found.push(key)
+        v = ctx.scope(key, () => f(ctx, j))
+        delete d[key]
```

Comment:
Could be useful to add a comment to indicate why it deletes keys as they are processed (same for field).